### PR TITLE
Fix un-authenticated behavior for all endpoints

### LINKED ISSUE #2193: Return 401 for unauthenticated requests to /api/ and redirect to login page for other urls
## Description
When user is not authenticated:
- Endpoints within /api/ should return a 401.
   - Currently some of them within /api/ return a 403, and some perform redirection to login page. They should all return a 401.
- All other endpoints which have associated views such as /, /<db_name>/, /<db_name>/<schema_id>/ etc., should redirect to login page.

## Frontend work
* Within api/requestUtils
  * Check if the response status is 401.
    * If yes, reload page.

### PR DESCRIPTION
**Pull request by Anish9901**
Fixes #2193

**Technical details**

This adds the DRF's `TokenAuthentication` authentication scheme that properly handles unauthenticated behavior by responding with `HTTP 401 Unauthorized` error when an unauthorized request is made to any `/api` endpoint.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: config/settings/common_settings.py
Comment (silentninja) [inline config/settings/common_settings.py:170]: `BasicAuthentication` has many [security vulnerabilities](https://www.whiteoaksecurity.com/blog/simple-security-fails-part-5-basic-authentication/), we should not be using it and I don't see why it is needed for to fix this issue

**Review by silentninja — CHANGES_REQUESTED** [review]
I don't think this PR actually fixes the issue, instead replaces it with a completely different behaviour compared to the expected outcome mentioned in issue #2193. When a user is not authenticated, the outcomes listed below happen randomly,
  - A browser prompt asking for a username and password
   - Sometimes results in a redirect to the login page for `/api` endpoints
    - Sometimes results in 401 status code
I guess it is due to the authentication type changes due to how DRF handles authentication(tries all the authentication backends) and how authentication backends are chosen by Django, but in short this trouble in my opinion and I would prefer to avoid it.
    
~I think issue #2193 is happening due to the [`AccessViewSetMixin` ](https://github.com/rsinger86/drf-access-policy/blob/8c4630432aafd70ff3fecb70d0606dc30426dbf8/rest_access_policy/access_view_set_mixin.py) as it overrides the permission classes and uses its access policy permission classes before other DRF permission classes, so when an unauthenticated user tries to access an API, it results in 403 status code(thrown by the `AccessViewSetMixin` when any condition fails). So we should try to change the priority of the permission classes so that the DRF [`isAuthenticated`](https://github.com/centerofci/mathesar/blob/f33c7a63d2a0e7b2bc7ce0f9eb35b0783b2b6ae6/config/settings/common_settings.py#L174) permission class(added by default to all viewset) is the top priority as it is throws `401` instead of `403`~

Edit: Looks like throwing 403 for session authentication is the [expected behaviour](https://www.django-rest-framework.org/api-guide/permissions/#how-permissions-are-determined). There is a long discussion (https://github.com/encode/django-rest-framework/pull/1611) regarding this problem.

**Comment by Anish9901** [comment]
> Sometimes results in a redirect to the login page for /api endpoints

@silentninja I didn't find any `/api` endpoint redirecting to the login page during local testing. Can you provide some endpoints that redirect to login page instead of 401s?

About the popups I thought that frontend could disable them as per the need. Correct me if I am wrong.

File: config/settings/common_settings.py
Comment (Anish9901) [reply config/settings/common_settings.py]: Replaced with `TokenAuthentication`.

I was unaware of the security vulnerabilities of `BasicAuthentication` as they weren't mentioned any where in the [docs](https://www.django-rest-framework.org/api-guide/authentication/#basicauthentication).

>  I don't see why it is needed for to fix this issue

`BasicAuthentication` did provide the appropriate `WWW-Authenticate` header which was necessary for returning `HTTP 401 Unauthorized` which was lacking in `SessionAuthentication` therefore I decided to go with it.

**Comment by silentninja** [comment]
> > Sometimes results in a redirect to the login page for /api endpoints
> 
>  I didn't find any `/api` endpoint redirecting to the login page during local testing. Can you provide some endpoints that redirect to login page instead of 401s?

This behaviour does not happen with a particular API, instead, it is one of the randomly chosen actions when an unauthenticated user tries to access an API. I don't know of a reliable way to reproduce it. But lets not worry about it now. The commit doesn't seem to have that issue.

**Review by silentninja — CHANGES_REQUESTED** [review]

File: mathesar/api/db/permissions/table.py
Comment (Anish9901) [reply mathesar/api/db/permissions/table.py:51]: I missed it, thanks for the catch!

**Review by silentninja — APPROVED** [review]

**Comment by silentninja** [comment]
Thanks for the PR @Anish9901, it is working as expected. I will wait for @pavish to review it before merging the PR

**Review by pavish — APPROVED** [review]
@Anish9901 Looks good. I tested with a couple endpoints and it works as expected, I did not do a through test since @silentninja has already reviewed this.

I added a frontend commit https://github.com/centerofci/mathesar/pull/2460/commits/16918496879ebe40870e24ce306b136619ad75a2 to handle reload (which redirects to login page) when a 401 is encountered.

**Pull request merged**
